File: tools/geo.py
"""座標の土台(metric-v1)。geo.js と同じ変換。定数は tools/data/frame.json が正。

mx,my = 原点からの東向き・南向きのメートル(北が上)。数値1 = 1m。
"""
import json
import math
import os
import re

_HERE = os.path.dirname(os.path.abspath(__file__))
_ROOT = os.path.dirname(_HERE)
FRAME = json.load(open(os.path.join(_HERE, 'data', 'frame.json')))
LAT0, LON0, OX, OY = FRAME['lat0'], FRAME['lon0'], FRAME['ox'], FRAME['oy']
KX = 111320 * math.cos(math.radians(LAT0))
KY = 110950


def ll2m(lat, lon):
    """緯度経度 → (mx, my)"""
    return ((lon - LON0) * KX + OX, -(lat - LAT0) * KY + OY)


def m2ll(mx, my):
    """(mx, my) → (lat, lon)"""
    return (LAT0 - (my - OY) / KY, LON0 + (mx - OX) / KX)


def _check_js_matches():
    """geo.js の定数が frame.json と食い違っていたら落とす(二重管理の事故防止)"""
    src = open(os.path.join(_ROOT, 'geo.js')).read()
    for key in ('lat0', 'lon0', 'ox', 'oy'):
        m = re.search(rf"\b{key}:\s*(-?[\d.]+)", src)
        assert m and abs(float(m.group(1)) - FRAME[key]) < 1e-9, f'geo.js の {key} が frame.json と不一致'
    m = re.search(r"name:\s*'([^']+)'", src)
    assert m and m.group(1) == FRAME['name'], 'geo.js の name が frame.json と不一致'


_check_js_matches()

File: geo.js
// 座標の土台(metric-v1)。tools/data/frame.json と同じ値を持つ(tools/geo.py が一致を検査する)
// mx,my = 原点からの東向き・南向きのメートル(北が上)。数値1 = 1m(小数可)
export const FRAME = {
  name: 'metric-v1',
  lat0: 34.702,
  lon0: 135.497,
  ox: 843.39,
  oy: 944.35,
};
const KX = 111320 * Math.cos(FRAME.lat0 * Math.PI / 180); // 1度の経度 ≒ この距離(m)
const KY = 110950;                                          // 1度の緯度 ≒ 110950m

// 緯度経度 → [mx, my]
export function ll2m(lat, lon) {
  return [(lon - FRAME.lon0) * KX + FRAME.ox, -(lat - FRAME.lat0) * KY + FRAME.oy];
}
// [mx, my] → [lat, lon]
export function m2ll(mx, my) {
  return [FRAME.lat0 - (my - FRAME.oy) / KY, FRAME.lon0 + (mx - FRAME.ox) / KX];
}
